Modernize the webserver code (part 1)

timvandermeij opened the pull request:
The commit messages contain more details about the individual changes. This first part mainly focuses on extracting self-contained functionality from the large `handler` method to make it more manageable and to get rid of using variables from the outer scope (with the end goal of enabling the `no-var` rule for this file).

_The first and last commits should be reviewed with the `?w=1` flag for an easier diff._

timvandermeij committed:
Convert the webserver to a proper class with private methods

timvandermeij committed:
Extract and modernize the webserver's file serving code

The `handler` method contained this code in an inline function, which
made the `handler` method big and harder to read. Moreover, this code
relied on variables from the outer scope, which made it harder to reason
about because the inputs and outputs weren't easily visible.

This commit fixes the problems by extracting the file serving code into
a dedicated private method, and modernizing it to use e.g. `const`/`let`
instead of `var` and using template strings.

timvandermeij committed:
Extract and modernize the webserver's range file serving code

The `handler` method contained this code in an inline function, which
made the `handler` method big and harder to read. Moreover, this code
relied on variables from the outer scope, which made it harder to reason
about because the inputs and outputs weren't easily visible.

This commit fixes the problems by extracting the range file serving code
into a dedicated private method, and modernizing it to use e.g. `const`/
`let` instead of `var` and using template strings.

timvandermeij committed:
Extract and modernize the webserver's directory listing code

The `handler` method contained this code in an inline function, which
made the `handler` method big and harder to read. Moreover, this code
relied on variables from the outer scope, which made it harder to reason
about because the inputs and outputs weren't easily visible.

This commit fixes the problems by extracting the directory listing code
into a dedicated private method, and modernizing it to use e.g. `const`/
`let` instead of `var` and using template strings.

Snuffleupagus commented on test/webserver.mjs:
```diff
@@ -188,154 +194,148 @@ WebServer.prototype = {
+    }
+
+    const all = queryPart === "all";
+    fs.readdir(directory, function (error, files) {
+      if (error) {
+        response.end();
```
Given that we require at least Node.js version 18, would it be feasible to make `#serveDirectoryIndex` async such that we can use the [promise-version](https://nodejs.org/api/fs.html#fspromisesreaddirpath-options) of the `fs` module instead?
That might lead to nicer code here, since we could then do something like below, which avoids the callback-structure and reduces indentation.
```suggestion
    let files;
    try {
      files = await fsPromises.readdir(directory);
    } catch {
      response.end();
      return;
    }
    response.write(
      ...
```

timvandermeij replied:
Good idea; done in the new commit.

timvandermeij commented:
The changes since the original commits are easiest to view with the interdiff in no-whitespace mode: https://github.com/mozilla/pdf.js/compare/af5764273b29b2e34ead0b54bfefe3ca912fedd1..935e8c8c5bec6cb1e36d8976780ca2d61e9118f2?w=1.

I made the following changes:

- Use `fsPromises.readdir` instead of `fs.readdir` and making the method asynchronous.
- Use `for/of` loop for the files instead of `files.forEach`.
- Moved `all` and `escapeHTML` down in the method to have the definition closer to where it's actually used.
- Introduced a couple of newlines to visually separate blocks of logic in the method for better readability.

Snuffleupagus commented on test/webserver.mjs:
```diff
@@ -158,7 +163,7 @@ WebServer.prototype = {
       }
       if (isDir) {
-        serveDirectoryIndex(filePath);
+        await self.#serveDirectoryIndex(res, pathPart, queryPart, filePath);
```
Do we actually need to await this call (and thus make `statFile` async), since we're not actually returning anything from this method?
Even before this patch the function was asynchronous, but in a different way, since `fs.readdir` used a callback function.

timvandermeij replied:
I have checked and it's indeed not necessary, so I have removed it in the new commit. Thanks!

Snuffleupagus reviewed (approved):
I've tested this locally and everything seems to work just as before.

r=me, thank you!

Snuffleupagus commented:
/botio test
